From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 24 May 2024 00:03:50 +0900
Subject: [PATCH] EntityLookup accessibleEntities

Use a synchronized block as data is only read rarely for debug reports

diff --git a/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java b/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
index 15ee41452992714108efe53b708b5a4e1da7c1ff..1020fdfe10a27dbb43f272d32b1366cf1e56fd8a 100644
--- a/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
+++ b/src/main/java/io/papermc/paper/chunk/system/entity/EntityLookup.java
@@ -191,11 +191,15 @@ public final class EntityLookup implements LevelEntityGetter<Entity> {
 
     @Override
     public Iterable<Entity> getAll() {
+        synchronized (this.accessibleEntities) { // MultiPaper
         return new ArrayIterable<>(this.accessibleEntities.getRawData(), 0, this.accessibleEntities.size());
+        } // MultiPaper
     }
 
     public Entity[] getAllCopy() {
+        synchronized (this.accessibleEntities) { // MultiPaper
         return Arrays.copyOf(this.accessibleEntities.getRawData(), this.accessibleEntities.size(), Entity[].class);
+        } // MultiPaper
     }
 
     @Override
@@ -277,7 +281,9 @@ public final class EntityLookup implements LevelEntityGetter<Entity> {
                     if (newVisibility.ordinal() > oldVisibility.ordinal()) {
                         // status upgrade
                         if (!oldVisibility.isAccessible() && newVisibility.isAccessible()) {
+                            synchronized (this.accessibleEntities) { // MultiPaper
                             this.accessibleEntities.add(entity);
+                            } // MultiPaper
                             EntityLookup.this.worldCallback.onTrackingStart(entity);
                         }
 
@@ -291,7 +297,9 @@ public final class EntityLookup implements LevelEntityGetter<Entity> {
                         }
 
                         if (oldVisibility.isAccessible() && !newVisibility.isAccessible()) {
+                            synchronized (this.accessibleEntities) { // MultiPaper
                             this.accessibleEntities.remove(entity);
+                            } // MultiPaper
                             EntityLookup.this.worldCallback.onTrackingEnd(entity);
                         }
                     }
